utils/zerovec/design_doc.md

```diff
@@ -0,0 +1,417 @@
+
+```rust
+struct Foo {
+    id: u32,
+    other_field: char,
+    value: String,
+}
+
+// Implements VarULE
+struct FooULE {
+    id: u32::ULE,
+    other_field: char::ULE,
+    value: str
+}
+```
+
+If we wish to push the data from a `Foo` to a `VarZeroVec`, we would have to first allocate a `Box<FooULE>`, pass it by-reference to the `VarZeroVec`, and then discard the allocation. This is highly inefficient.
+
+To help perform such mutations, the [`EncodeAsVarULE`] trait exists:
+
+
+```rust
+pub unsafe trait EncodeAsVarULE<T: VarULE + ?Sized> {
+    fn encode_var_ule_as_slices<R>(&self, cb: impl FnOnce(&[&[u8]]) -> R) -> R;
+
+    fn encode_var_ule_len(&self) -> usize { ... }
+    fn encode_var_ule_write(&self, dst: &mut [u8]) { ... }
+}
+```
+
+Users must either implement `encode_var_ule_as_slices` or the other two methods, everything else will be handled automatically. These methods allow users to take arbitrary types and specify how they can be encoded into a byte stream to produce a valid `T: VarULE` type.
+
+
+We plan to have a proc macro for implementing this trait.
+
+
+### ZeroMap traits
+
+There's a suite of traits that enables `ZeroMap` to work, largely abstracting over `ZeroVec` and `VarZeroVec`.
+
+The core trait is [`ZeroMapKV`], which is the only trait users need to implement. It essentially directs `ZeroMap` to select the appropriate container type for a given `Self` key/value type.
+
+```rust
+pub trait ZeroMapKV<'a> {
+    type Container: MutableZeroVecLike<'a, Self, GetType = Self::GetType, OwnedType = Self::OwnedType> + Sized;
+    type GetType: ?Sized + 'static;
+    type OwnedType: 'static;
+}
+```
+
+Implementing it is pretty straightforward. If you are implementing it on an `AsULE` type, the implementation will look like:
+
+```rust
+impl<'a> ZeroMapKV<'a> for u32 {
+    type Container = ZeroVec<'a, u32>;
+    type GetType = u32::ULE;
+    type OwnedType = u32;
+}
+```
+
+and for `VarULE` types, it will look like:
+
+```rust
+impl<'a> ZeroMapKV<'a> for str {
+    type Container = VarZeroVec<'a, str>;
+    type GetType = str;
+    type OwnedType = Box<str>;
+}
+```
+
+Aside from this trait, there are traits [`MutableZeroVecLike`], [`BorrowedZeroVecLike`], and [`ZeroVecLike`], which abstract over zero-copy vector collections that are mutable, pure-borrowed, and general respectively.
+
+
+## Proc macros
+
+It's important for users to be able to use their own types within `ZeroVec` and `VarZeroVec`.
+
+To aid this, we can add a couple proc macros.
+
+
+### `#[derive(ULE)]` and `#[derive(VarULE)]`
+
+These are basic derives that can be applied to types to _just_ generate ULE and VarULE implementations for types whose definition is _already_ ULE compatible.
+
+These can only be applied to structs where all fields are ULE types (for `#[derive(VarULE)]`, the last field must be an unsized `VarULE` type). These derives will do the following things:
+
+ - Apply `#[repr(packed)]` to the type
```

Comment:
If all fields are `ULE` then it doesn't need `#[repr(packed)]`, right?

Reply:
No, it still needs it

Reply:
Does repr(C) produce the same result as repr(packed) on types with ULE fields?

Reply:
I'm not sure.... _maybe_? It's worth looking in to. It would certainly make some things easier since rust has restrictions on packed types.